[SPARK-29861][CORE] Reduce downtime in Spark standalone HA master switch

Nibooor opened the pull request:
### What changes were proposed in this pull request?
1. Close open zookeeper connections during spark shutdown (SIGTERM)
2. Bump the curator version to 4.2.0
3. Implement a custom error policy that is tolerant to a zookeeper connection suspension.

### Why are the changes needed?
As described in [SPARK-29861](https://issues.apache.org/jira/browse/SPARK-29861), the recovery process of Spark (standalone) master in HA with zookeeper took about 1-2 minutes. During this time no spark master is active, which makes interaction with spark essentially impossible.

Justification for the changes:
1. because the spark master election does not properly restart if there are still open zookeeper connections
2. as a precondition for 3
3. to not restart spark when the connection is suspended temporarily, but only if it is really closed.

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
* Set up spark with two masters in standalone mode and zookeeper
* Kill active master
* Observe that the second master takes over immediately (within a few milliseconds)

sapbinw committed:
Close zookeeper connections on SIGTERM

On spark shutdown the curator client connections are not closed.
This will unnecessarily slow down the standby master switch in
standalone mode, as zookeeper ephemeral nodes will not be deleted
with the session still open.
Therefore the zookeeper leader election will only be started after
the session of a terminated master is expired.
Fix this by registering a signal handler to SIGTERM, which properly
closes the curator connection in two places.

Change-Id: I7fa60b7f5ba7687d4021cbfe635e437fe05b4a17

sapbinw committed:
Add ZookeeperSuspensionTolerantErrorPolicy

* Bump curator version to 4.2.0
* Implement ZookeeperSuspensionTolerantErrorPolicy that treats
  only LOST (and not SUSPENDED) connections as error states.
* Init CuratorFramework with ZookeeperSuspensionTolerantErrorPolicy
  in SparkCuratorUtil
* Exclude curator from hadoop for compatibility reasons
* Use curator-test 2.12.0 for compatibility reasons

A suspended connection previously lead to a spark restart. This
lead to unnecessarily restarts of spark masters, that effectively
prolonged the HA master switch in standalone mode.

Change-Id: I01bd76cca499d03cb114b7d9ddfb797f2033f3cb

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/deploy/master/ZookeeperSuspensionTolerantErrorPolicy.scala:
```diff
@@ -0,0 +1,32 @@
+package org.apache.spark.deploy.master
+
+import org.apache.curator.framework.state.ConnectionState
+import org.apache.curator.framework.state.ConnectionStateErrorPolicy
```
Hi. Thank you for making a PR.
However, GitHub Action shows the compilation error. Is this compiled in your environment?
```
[ERROR] [Error] /home/runner/work/spark/spark/core/src/main/scala/org/apache/spark/deploy/master/ZookeeperSuspensionTolerantErrorPolicy.scala:21: object ConnectionStateErrorPolicy is not a member of package org.apache.curator.framework.state
318
[ERROR] [Error] /home/runner/work/spark/spark/core/src/main/scala/org/apache/spark/deploy/master/ZookeeperSuspensionTolerantErrorPolicy.scala:26: not found: type ConnectionStateErrorPolicy
319
[ERROR] [Error] /home/runner/work/spark/spark/core/src/main/scala/org/apache/spark/deploy/SparkCuratorUtil.scala:47: value connectionStateErrorPolicy is not a member of org.apache.curator.framework.CuratorFrameworkFactory.Builder
```

Nibooor replied:
Hi, 

yes this happens in my local environment now as well (I cherry-picked this from a 2.2 branch before).

The issue is caused because according to the build logs `curator-client-2.13.0.jar` is added to the core classpath by `maven-dependency-plugin:3.1.1` instead of the desired version `4.2.0`.

Strangely `./build/mvn dependency:tree` lists only the correct versions (`4.2.0`) on curator everywhere...

Nibooor replied:
Oversaw a different version in curator that was not present on the previous branch.
Compiles now.

sapbinw committed:
Remove explicit curator import in hadoop-3.2 profile

It overwrites the hadoop exclusion rule of curator (with a more
recent version) otherwise.

Change-Id: If50cdd37c5381d563566c728d1aaf1e67c45faba

sapbinw committed:
Adapt dependency changes to profile 'hive-1.2/2.3'

The profile was recently added and did not reflect the proposed
changes in the dependency file yet.

Change-Id: I982048d91ebe8cb81cdb50c5d857ed44e261b5c3

dongjoon-hyun commented:
ok to test

dongjoon-hyun reviewed (commented):
BTW, `Bump the curator version to 4.2.0` is required? If not, please remove that dependency upgrade.

SparkQA commented:
**[Test build #115347 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115347/testReport)** for PR 26598 at commit [`421ddcc`](https://github.com/apache/spark/commit/421ddccd4299ef601fee0c1ca03e6167997ca4f7).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20153/
Test PASSed.

SparkQA commented:
**[Test build #115347 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115347/testReport)** for PR 26598 at commit [`421ddcc`](https://github.com/apache/spark/commit/421ddccd4299ef601fee0c1ca03e6167997ca4f7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115347/
Test FAILed.

Nibooor commented:
> BTW, Bump the curator version to 4.2.0 is required? If not, please remove that dependency upgrade.
Yes, sadly this is part of the improvement.

The custom error policy was [introduced starting curator 3.0](https://github.com/apache/curator/pull/98), so we need a bump anyways. I decided to do that to the highest stable version if this needs a bump anyways.

I am looking into the failed test now.

Nibooor commented:
SBT misses the class `ConnectionHandlingPolicy` during test runtime and therefore has one test case failing. The class is introduced in curator 3 (so should be available with the version bump). 

I've had the same error locally before because of a dependency eviction and solved it introducing the exclusion for curator in the pom (at the hadoop depedency).
Now the tests succeeds for me locally in maven, but apparently not in sbt.

Is it possible the pom dependency exclusion does not get translated properly into sbt? My ivy report shows that in the phase `test-internal` the dependency is evicted by an old version in favor of hadoop 2.7.4.
Or do I have to add some exclusion rule equivalent in sbt to make this work there as well?
